### PR TITLE
Upgrades: New - Admin tool to re-migrate globals to kit.

### PR DESCRIPTION
**Pull request by matipojo**

### CODE REVIEW

File: core/upgrade/elementor-3-re-migrate-globals.php [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+
+	private function notice( $message ) {
+		$logger = Plugin::$instance->logger->get_logger();
+		$logger->notice( $message );
+	}
+
+	public function __construct() {
+		add_action( 'elementor/admin/after_create_settings/' . Tools::PAGE_ID, [ $this, 'register_admin_tools_fields' ], 60 /* After plugins rollback */ );
+		add_action( 'elementor/admin/localize_settings', [ $this, 'admin_localize_settings' ] );
+		add_action( 'wp_ajax_elementor_re_migrate_globals', [ $this, 'ajax_elementor_re_migrate_globals' ] );
```
Comment (kobizz) [inline core/upgrade/elementor-3-re-migrate-globals.php:159]: Use Ajax Manager

File: assets/dev/js/admin/admin.js [hunk 2]
```diff
@@ -242,6 +242,32 @@
...
+					headerMessage: self.translate( 're_migrate_globals' ),
+					message: self.translate( 're_migrate_globals_confirm' ),
+					strings: {
+						confirm: self.translate( 'yes' ),
+						cancel: self.translate( 'cancel' ),
+					},
+					onConfirm: function() {
+						$this.removeClass( 'success' ).addClass( 'loading' );
+
+						$.post( ajaxurl, {
```
Comment (kobizz) [inline assets/dev/js/admin/admin.js:260]: `elementorCommon.ajax.addRequest()`

File: assets/dev/js/admin/admin.js [hunk 3]
```diff
@@ -242,6 +242,32 @@
 				} ).show();
 			} );
 
+			$( '.elementor-re-migrate-globals-button' ).on( 'click', function( event ) {
```
Comment (kobizz) [inline assets/dev/js/admin/admin.js:245]: 1. At least for newly added code, use caching
2. Use arrow function

File: assets/dev/js/admin/admin.js [hunk 4]
```diff
@@ -242,6 +242,32 @@
...
+				var $this = $( this );
+
+				elementorCommon.dialogsManager.createWidget( 'confirm', {
+					headerMessage: self.translate( 're_migrate_globals' ),
+					message: self.translate( 're_migrate_globals_confirm' ),
+					strings: {
+						confirm: self.translate( 'yes' ),
+						cancel: self.translate( 'cancel' ),
+					},
+					onConfirm: function() {
```
Comment (kobizz) [inline assets/dev/js/admin/admin.js:257]: Arrow function

File: assets/dev/js/admin/admin.js [hunk 5]
```diff
@@ -242,6 +242,32 @@
...
+						$this.removeClass( 'success' ).addClass( 'loading' );
+
+						$.post( ajaxurl, {
+							action: 'elementor_re_migrate_globals',
+							_nonce: $this.data( 'nonce' ),
+						} )
+						.done( function() {
+							$this.removeClass( 'loading' ).addClass( 'success' );
```
Comment (kobizz) [inline assets/dev/js/admin/admin.js:265]: Arrow function

File: core/upgrade/elementor-3-re-migrate-globals.php [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+
+	private function run() {
+		$this->notice( 'Re-migrate 3.0.0 globals: Start' );
+
+		$callbacks = [];
+
+		$active_kit_id = Plugin::$instance->kits_manager->get_active_id();
+		$kit = Plugin::$instance->documents->get( $active_kit_id );
+		// Already exist. use raw settings that doesn't have default values.
+		$meta_key = \Elementor\Core\Settings\Page\Manager::META_KEY;
```
Comment (kobizz) [inline core/upgrade/elementor-3-re-migrate-globals.php:76]: Create an alias

File: assets/dev/js/admin/admin.js [hunk 7]
```diff
@@ -242,6 +242,32 @@
 				} ).show();
 			} );
 
+			$( '.elementor-re-migrate-globals-button' ).on( 'click', function( event ) {
+				event.preventDefault();
+
+				var $this = $( this );
```
Comment (kobizz) [inline assets/dev/js/admin/admin.js:248]: const

**Review by kobizz — CHANGES_REQUESTED** [review]

File: core/upgrade/elementor-3-re-migrate-globals.php [hunk 8]
```diff
@@ -0,0 +1,161 @@
...
+		// Add the button to the rollback screen.
+		$tools->add_fields( 'versions', 'rollback', [
+			're_migrate_globals_separator' => [
+				'field_args' => [
+					'type' => 'raw_html',
+					'html' => '<hr>',
+				],
+			],
+			're_migrate_globals' => [
+				'label' => __( 'Rerun Update Script', 'elementor-pro' ),
```
Comment (kobizz) [inline core/upgrade/elementor-3-re-migrate-globals.php:54]:
```suggestion
				'label' => __( 'Rerun Update Script', 'elementor' ),
```

File: core/upgrade/elementor-3-re-migrate-globals.php [hunk 9]
```diff
@@ -0,0 +1,161 @@
...
+				'field_args' => [
+					'type' => 'raw_html',
+					'html' => sprintf(
+						'<a href="#" data-nonce="%s" class="button elementor-button-spinner elementor-re-migrate-globals-button">%s</a>',
+						wp_create_nonce( 'elementor_re_migrate_globals' ),
+						__( 'Migrate to v3.0', 'elementor-pro' )
```
Comment (kobizz) [inline core/upgrade/elementor-3-re-migrate-globals.php:60]:
```suggestion
						__( 'Migrate to v3.0', 'elementor' )
```

**Review by kobizz — CHANGES_REQUESTED** [review]

File: core/upgrade/elementor-3-re-migrate-globals.php (hunk 1 above)
Comment (matipojo) [reply core/upgrade/elementor-3-re-migrate-globals.php:159]: Done.

File: assets/dev/js/admin/admin.js (hunk 2 above)
Comment (matipojo) [reply assets/dev/js/admin/admin.js:260]: Done.

File: assets/dev/js/admin/admin.js (hunk 3 above)
Comment (matipojo) [reply assets/dev/js/admin/admin.js:245]: 1. Done.
2. The `this` should be referred to the event target.

File: assets/dev/js/admin/admin.js (hunk 4 above)
Comment (matipojo) [reply assets/dev/js/admin/admin.js:257]: Done.

File: assets/dev/js/admin/admin.js (hunk 5 above)
Comment (matipojo) [reply assets/dev/js/admin/admin.js:265]: Done.

File: core/upgrade/elementor-3-re-migrate-globals.php (hunk 6 above)
Comment (matipojo) [reply core/upgrade/elementor-3-re-migrate-globals.php:76]: Done.

File: assets/dev/js/admin/admin.js (hunk 7 above)
Comment (matipojo) [reply assets/dev/js/admin/admin.js:248]: Done

File: core/upgrade/elementor-3-re-migrate-globals.php (hunk 8 above)
Comment (matipojo) [reply core/upgrade/elementor-3-re-migrate-globals.php:54]: Done.

File: core/upgrade/elementor-3-re-migrate-globals.php (hunk 9 above)
Comment (matipojo) [reply core/upgrade/elementor-3-re-migrate-globals.php:60]: Done.

File: assets/dev/js/admin/admin.js (hunk 3 above)
Comment (kobizz) [reply assets/dev/js/admin/admin.js]: You can use `event.currentTarget` to achieve this

Comment (matipojo) [reply assets/dev/js/admin/admin.js]: Done.

**Review by kobizz — APPROVED** [review]

**Pull request merged**